In [10]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [11]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df = pd.read_csv(Path("lending_data.csv"))

# Review the DataFrame
print(df.head())

   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0  


### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [12]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df["loan_status"]

# Separate the X variable, the features
X = df.drop(columns="loan_status")

In [13]:
# Review the y variable Series
print(y.head())

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64


In [14]:
# Review the X variable DataFrame
print(X.head())

   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000  


### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [15]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [16]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model = LogisticRegression(random_state=1)

# Fit the model using training data
model.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [17]:
# Make a prediction using the testing data
y_pred = model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [18]:
# Generate a confusion matrix for the model
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(conf_matrix)

Confusion Matrix:
[[18663   102]
 [   56   563]]


In [19]:
# Print the classification report for the model
class_report = classification_report(y_test, y_pred)
print("Classification Report:")
print(class_report)

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** WRITE YOUR ANSWER HERE!

The logistic regression model performs exceptionally well in predicting both the 0 (healthy loan) and 1 (high-risk loan) labels, as evidenced by the classification report. Here's a detailed analysis:

***Precision***

For healthy loans (label 0), the precision is 1.00. This means that every loan predicted as healthy by the model is indeed healthy, indicating no false positives.
For high-risk loans (label 1), the precision is 0.85. This suggests that 85% of loans predicted as high-risk are actually high-risk, with some false positives.

***Recall***

For healthy loans (label 0), the recall is 0.99. This means that the model correctly identifies 99% of all actual healthy loans, missing only 1% (false negatives).
For high-risk loans (label 1), the recall is 0.91. This means that the model correctly identifies 91% of all actual high-risk loans, missing 9% (false negatives).
F1-Score:

The F1-score for healthy loans (label 0) is 1.00, which is the best possible score, indicating a perfect balance between precision and recall.
The F1-score for high-risk loans (label 1) is 0.88, which is quite high and indicates a good balance between precision and recall.

***Support***

The support shows the number of instances for each class in the dataset: 18,765 healthy loans and 619 high-risk loans.

***Overall Accuracy***

The overall accuracy of the model is 0.99, meaning the model correctly predicts the loan status for 99% of the instances in the test set.

***Macro Average***

The macro average precision, recall, and F1-score are 0.92, 0.95, and 0.94 respectively. These averages treat both classes equally without considering the class imbalance.

***Weighted Average***

The weighted average precision, recall, and F1-score are 0.99, 0.99, and 0.99 respectively. These averages account for the class imbalance by giving more weight to the majority class (healthy loans).

***Conclusion***
The logistic regression model predicts healthy loans (label 0) with near-perfect accuracy, precision, recall, and F1-score. For high-risk loans (label 1), the model also performs well, with an F1-score of 0.88, indicating a good balance between precision and recall. The overall accuracy of 0.99 shows that the model is highly effective in predicting loan statuses. Despite the class imbalance, where healthy loans significantly outnumber high-risk loans, the model maintains high performance across all metrics, demonstrating its reliability and effectiveness in distinguishing between healthy and high-risk loans.